atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/ReloadingCqlClusterContainer.java

```diff
@@ -0,0 +1,147 @@
+
+        this.refreshableSubscriptionDisposable = refreshableCqlCluster.subscribe(cqlCluster -> {
+            Optional<CqlCluster> maybeCqlClusterToClose = currentCqlCluster.getAndSet(Optional.of(cqlCluster));
+            if (maybeCqlClusterToClose.isPresent()) {
+                try {
+                    maybeCqlClusterToClose.get().close();
+                } catch (IOException e) {
+                    log.warn("Failed to close CQL Cluster after reloading server list", e);
+                }
+            }
+        });
+    }
+
+    public static ReloadingCqlClusterContainer of(
+            CassandraClusterConfig cassandraClusterConfig,
+            Refreshable<CassandraServersConfig> refreshableCassandraServersConfig,
+            Namespace namespace) {
+        return of(cassandraClusterConfig, refreshableCassandraServersConfig, namespace, CqlCluster::create);
+    }
+
+    public static ReloadingCqlClusterContainer of(
+            CassandraClusterConfig cassandraClusterConfig,
+            Refreshable<CassandraServersConfig> refreshableCassandraServersConfig,
+            Namespace namespace,
+            CqlClusterFactory cqlClusterFactory) {
+        return new ReloadingCqlClusterContainer(
+                cassandraClusterConfig, refreshableCassandraServersConfig, namespace, cqlClusterFactory);
+    }
+
+    /**
+     * Synchronized: See {@link #close()}.
+     */
+    private synchronized CqlCluster createNewCluster(
+            CassandraClusterConfig cassandraClusterConfig,
+            CassandraServersConfig cassandraServersConfig,
+            Namespace namespace,
+            CqlClusterFactory cqlClusterFactory) {
+        if (isClosed) {
+            throw new SafeIllegalStateException(
+                    "Attempted to create a new cluster after the container was closed. If this happens repeatedly,"
+                            + " this is likely a bug in closing the container. Otherwise, it is highly likely that the"
+                            + " container was closed at the same time as the server list was updated. If so, this error"
+                            + " can be ignored.",
+                    UnsafeArg.of("keyspace", namespace));
+        }
+        return cqlClusterFactory.create(cassandraClusterConfig, cassandraServersConfig, namespace);
+    }
+
+    /**
+     * Synchronized: A lock is taken out to ensure no new CQL Clusters are created after retrieving the current stored
+     * cql cluster to close. By doing so, we avoid closing a cluster and subsequently creating a new one that is
+     * never closed.
+     */
+    @Override
+    public synchronized void close() throws IOException {
+        isClosed = true;
+        refreshableSubscriptionDisposable.dispose();
+        Optional<CqlCluster> maybeCqlClusterToClose = currentCqlCluster.get();
+        if (maybeCqlClusterToClose.isPresent()) {
+            maybeCqlClusterToClose.get().close();
```

Comment:
what if this throws? I've noticed that this is the same pattern as before except without the try, and without setting the new cql cluster. Again, is this something we want to encapsulate? maybe

```java
shutdownPreviousCluster(Optional<CqlCluster> maybeClusterToClose) {
...
}
```

Reply:
Refactored into a separate method. If there's an exception, then the only caller of close (for now?) (which is`CassandraRepairhelper#onRemoval`) logs the exception and drops it. In the general case, it should be handled however the underlying `Cluster#close` exception would otherwise be handled.